pkg/ddl/backfilling.go

```diff
@@ -675,7 +676,60 @@ func (dc *ddlCtx) runAddIndexInLocalIngestMode(
 		return errors.Trace(err)
 	}
 	job := reorgInfo.Job
-	opCtx := NewLocalOperatorCtx(ctx, job.ID)
+	opCtx, cancel := NewLocalOperatorCtx(ctx, job.ID)
+
+	// spawn a goroutine to check if the job is cancelled
+	// TODO: design a more systematic way to cancel DDL jobs
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		sctx, err := sessPool.Get()
+		if err != nil {
+			logutil2.Logger(opCtx).Warn(
+				"failed to get session for job canceling checker, job can still be canceled after ingest is done",
+				zap.Error(err))
+			return
+		}
+		defer sessPool.Put(sctx)
+
+		ticker := time.NewTicker(time.Second)
```

Comment:
maybe reduce reorgWaitTime, we will check it in worker's runJobStep loop anyway

https://github.com/pingcap/tidb/blob/9dff38ba98405422cb0eb15993f385efe9068b47/pkg/ddl/reorg.go#L403-L404